### ISSUE TITLE
Argon and Bcrypt hasher classes do not verify hashing algorithm

### ISSUE THREAD

**Issue by ioluas** [laravel/framework#24162]
- Laravel Version: 5.6.20
- PHP Version: 7.2.4
- Database Driver & Version: MySQL 5.7.22

### Description:
`check` method in both `ArgonHasher` & `BcryptHasher` do not verify hashing algorithm. The method only verifies that `$hashedValue` is not an empty string and then proceed to call native php function `password_verify`. Now, since the algorithm and salt are contained in `$hashedValue`, the return value from `password_verify` will not fail if the algorithm used is different than the one intended in the class. 
I had used `bcrypt` function in User model factory in a Laravel project, since the application is still under development; when I upgraded to 5.6, I switched hash driver to argon but forgot to update the factory and later realised I was still able to login with users credentials created with 'broken' factory.

### Steps To Reproduce:
Test method to clarify what I mean:
```php
public function testBasicVerification()
{
    $argonHasher = new \Illuminate\Hashing\ArgonHasher;
    $bcryptHasher = new \Illuminate\Hashing\BcryptHasher;
    $argonHashed = $argonHasher->make('password');
    $bcryptHashed = $bcryptHasher->make('password');
    $this->assertFalse($bcryptHasher->check('password', $argonHashed));
    $this->assertFalse($argonHasher->check('password', $bcryptHashed));
}
```
I'm not sure if you would agree this is a bug and I'm not sure I can think of a 'normal' scenario where this could happen. If I didn't forget to update the factory, I would have never looked at this twice, but on a second thought I think since the method exists in both classes, it makes sense for ArgonHasher to return false when asked to verify a Bcrypt hashed value and vice versa. 

This is a straight forward fix of adding an additional check on `password_get_info($hashedValue)['algo']` value in both methods.

**Comment by tillkruss** [comment]
Seems reasonable, can you submit a PR that throws an exception if the algorithms don't match? Should go to 5.7.

**Comment by ioluas** [comment]
Submitted PR #24170

**Comment by tillkruss** [comment]
Cool, make sure you submit a clean PR that can be merged.

**Comment by ioluas** [comment]
Sorry, forgive the rookie mistakes. submitted #24178

**Comment by tillkruss** [comment]
No worries, it's just a learning curve. PR looks good.

**Comment by khanks** [comment]
I don't know what kind of websites you guys run but you seem awfully cavalier about a "dot" update that breaks password hashing. There's no way that would be acceptable in my site.

**Comment by browner12** [comment]
Laravel does not follow semantic versioning.

https://laravel.com/docs/5.7/releases#versioning-scheme

This change was submitted and merged in master (5.7) before it was released. Breaking changes are allowed from 5.6 to 5.7
